crates/egui/src/widgets/text_edit/builder.rs

```diff
@@ -1050,7 +1050,8 @@ fn events(
             }
 
             Event::CompositionEnd(prediction) => {
-                if prediction != "\n" && prediction != "\r" && state.has_ime {
+                // CompositionEnd only characters may be typed into TextEdit without trigger CompositionStart first, so do not check ```state.has_ime = true``` in the following statement.
```

Comment:
```suggestion
                // CompositionEnd only characters may be typed into TextEdit without trigger CompositionStart first, so do not check `state.has_ime = true` in the following statement.
```

Is it still correct to check it for `Event::CompositionUpdate` above?

Reply:
[This page](https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event) show that `compositionstart` could be fired after a user starts entering a Chinese character using a [Pinyin](https://en.wikipedia.org/wiki/Pinyin), so when entering some word except `compositionend` character, `componentstart` fires and `state.has_ime` will always be `true`, and `compositionupdate` will work well. so it's still correct to check it for `compositionupdate ` because `state.has_ime` is always `true` when entring word.

I have test the situation that remove all `state.has_ime` in `CompositionStart`, `CompositionUpdate`,`CompositionEnd`, it works well.

Also, `state.has_ime` is a variable that indicating current input state, maybe some other features will use this variable, so i think keep this variable is a good choice, although it seems useless at current time.